### PR TITLE
Show solution descriptions in the Word report

### PR DESCRIPTION
**Pull request by reinhardt**
except for France and Italy when measures in place are active

syslabcom/scrum#1245

### CODE REVIEW

File: src/euphorie/client/docx/compiler.py [hunk 1]
```diff
@@ -447,7 +454,20 @@ def add_measure(self, doc, heading, measure, implemented=False):
             ]
 
         m = measure
-        values = [_simple_breaks(m.action or "")]
+        action = m.action
```
Comment (ale-rt) [inline src/euphorie/client/docx/compiler.py:457]: To reduce the lines of code of this method and to save an `if` in an already complex method, what about having something like:
```
action = self.measure2action(measure)
```

On the base class `measure2action` returns `measure.action` but for the others it returns the code in the if block.

I hope you get the point.

File: src/euphorie/client/docx/views.py [hunk 2]
```diff
@@ -111,23 +111,32 @@ def get_risks_for(self, sql_module):
...
             risk_description = _sanitize_html(risk_description)
 
-            measures = [
-                measure.action
-                for measure in (
-                    sql_risk.in_place_standard_measures
-                    + sql_risk.in_place_custom_measures
-                )
-            ]
+            # XXX can we move this distinction to the compiler?
```
Comment (ale-rt) [inline src/euphorie/client/docx/views.py:114]: That seems a good idea

File: src/euphorie/client/docx/views.py [hunk 3]
```diff
@@ -111,23 +111,32 @@ def get_risks_for(self, sql_module):
...
-                    _get_action_plan(action)
-                    for action in (
-                        sql_risk.standard_measures + sql_risk.custom_measures
+                    _get_action_plan(
+                        action, solution=risk.get(action.solution_id, None)
                     )
+                    for action in (sql_risk.standard_measures)
                 ]
+                actions.extend(
+                    [_get_action_plan(action) for action in (sql_risk.custom_measures)]
```
Comment (ale-rt) [inline src/euphorie/client/docx/views.py:138]:
```suggestion
                    [_get_action_plan(action) for action in sql_risk.custom_measures]
```

File: src/euphorie/client/docx/views.py [hunk 4]
```diff
@@ -111,23 +111,32 @@ def get_risks_for(self, sql_module):
...
                 risk and getattr(risk, "type", None) == "top5"
             ):
                 actions = [
-                    _get_action_plan(action)
-                    for action in (
-                        sql_risk.standard_measures + sql_risk.custom_measures
+                    _get_action_plan(
+                        action, solution=risk.get(action.solution_id, None)
                     )
+                    for action in (sql_risk.standard_measures)
```
Comment (ale-rt) [inline src/euphorie/client/docx/views.py:135]:
```suggestion
                    for action in sql_risk.standard_measures
```

**Review by ale-rt — APPROVED** [review]
LGTM and I am ok to merge this one.

It would be cool if you could check my suggestions.

Generally speaking, I think it is a good practice to not introduce feature flags and if all over the place to handle custom behaviors because it often leads to code which is harder to maintain.

If you decide not to rework this PR, please bear in mind for the next time.

**Comment by reinhardt** [comment]
Thanks for the suggestions. I see even more potential to improve this code. I'll check how much effort we can spend.

**Comment by reinhardt** [comment]
> Generally speaking, I think it is a good practice to not introduce feature flags and if all over the place to handle custom behaviors because it often leads to code which is harder to maintain.

What would you say is a better alternative?

**Comment by ale-rt** [comment]
> > Generally speaking, I think it is a good practice to not introduce feature flags and if all over the place to handle custom behaviors because it often leads to code which is harder to maintain.
> 
> What would you say is a better alternative?

See my suggestion in https://github.com/euphorie/Euphorie/pull/596#discussion_r1221339895

You code looks like

```python

class A:
    flag_active = False

    def __call__(self):
        if self.flag_active:
            return "something"
        else:
            return "something else"

class B(A):
    flag_active = True
```

My suggestion is to make it look like to this:

```python

class A:
    def __call__(self):
        return "something"

class B(A):
    def __call__(self):
        return "something else"
```

That's especially helpful in this case because of the cyclomatic complexity of the method.

**Comment by reinhardt** [comment]
Oh, that's what you mean. Now I get it, thanks. I do see the benefits. On the other hand, if I have a flag with a speaking name like `use_solution_descriptions`,  it is potentially clearer what the intention is.

**Comment by reinhardt** [comment]
I'll merge this now and clean up in a new pull request.

**Pull request merged**